### DISCUSSION TITLE
Macros

### DISCUSSION THREAD

**Discussion by techouse** [lejard-h/chopper#620] [Ideas]
With the advent of [Dart Macros](https://dart.dev/language/macros) we should probably start looking into adopting them in Chopper v9.

What are your thoughts, guys?

CC / @JEuler @Guldem

**Comment by Guldem** [comment]
I'm all for it :D I have been following the macros development for a year or so now and excited its almost here. 

Also I've already done some tests with it, exploring and possible chopper route. Currently there is still something I haven't figured out. That is obtaining meta data in macros. For example when annotating a parameter in a method with `Body(), Param(), etc.`. This is the only actual information I was able to find regarding this: https://github.com/dart-lang/sdk/blob/6bc417dd17f5edb228e60b9aebb70650bf7d5f50/tests/language/macros/json/json_serializable.dart#L275

But issues can be solved 😉 How should we go further with this. Include it in a separated package or in the chopper package itself? 
Do we want to split some work? And will it impact existing features?

**Reply by Guldem** [reply]
After some more exploring I have found that I see metadata when using `VSCode` `Go to Augment`. WHen using intellij or cli I still don't see the same result. Not sure why or how but learning more and more 😉

**Comment by Guldem** [comment]
I'm also always thinking on how to improve it and implement it the best it could be. Currently chopper is based on Retrofit and build with what build_runner is capable of. But macros and augmentations might offer new and other ways of define a chopper client.

For example:
```dart
@Get(uri: 'awesome/{id}',returnType: MyClass)
@ChopperClient('http://example.com')
class MyClient {
}

void main() async {
final client = MyClient();

final myClass = await client.getMyClass(id: '1');
}
```

I think we should not just port chopper to macros but also explore the best way this could work with macros.

**Reply by y3l1n4ung** [reply]
If we use this method ```@Get(uri: 'awesome/{id}',returnType: MyClass)```
I think We don't need to worry about `@Body @Param` annotation , just add  body , param, path  into params of injected method. 😃

**Reply by Guldem** [reply]
That was also why I came up with this. But it would change the way your used to how its working now.

**Comment by Guldem** [comment]
Well I guess this will only be a dream for now:
https://medium.com/dartlang/an-update-on-dart-macros-data-serialization-06d3037d4f12

**Reply by techouse** [reply]
Sadly. I'm gonna close down this convo then.
